cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -4282,40 +4282,115 @@
       )
        (require (or (= ty 32) (= ty 64)))
 )
-(rule (x64_add_mem $I8 addr val) (x64_addb_mr_mem addr val))
-(rule (x64_add_mem $I16 addr val) (x64_addw_mr_mem addr val))
-(rule (x64_add_mem $I32 addr val) (x64_addl_mr_mem addr val))
-(rule (x64_add_mem $I64 addr val) (x64_addq_mr_mem addr val))
-
-(decl x64_sub_mem (Type Amode Gpr) SideEffectNoResult)
-(rule (x64_sub_mem $I8 addr val) (x64_subb_mr_mem addr val))
-(rule (x64_sub_mem $I16 addr val) (x64_subw_mr_mem addr val))
-(rule (x64_sub_mem $I32 addr val) (x64_subl_mr_mem addr val))
-(rule (x64_sub_mem $I64 addr val) (x64_subq_mr_mem addr val))
-
-(decl x64_and_mem (Type Amode Gpr) SideEffectNoResult)
+
+;; `add mem, reg`
+(rule 0 (x64_add_mem $I8 addr val) (x64_addb_mr_mem addr val))
+(rule 0 (x64_add_mem $I16 addr val) (x64_addw_mr_mem addr val))
+(rule 0 (x64_add_mem $I32 addr val) (x64_addl_mr_mem addr val))
+(rule 0 (x64_add_mem $I64 addr val) (x64_addq_mr_mem addr val))
+
+;; `add mem, imm`
+(rule 1 (x64_add_mem $I8 addr (u8_from_iconst val)) (x64_addb_mi_mem addr val))
+(rule 1 (x64_add_mem $I16 addr (u16_from_iconst val)) (x64_addw_mi_mem addr val))
+(rule 1 (x64_add_mem $I32 addr (u32_from_iconst val)) (x64_addl_mi_mem addr val))
+(rule 1 (x64_add_mem $I64 addr (i64_from_iconst val))
+      (if-let val2 (i32_try_from_i64 val))
```

Comment:
This could replace `i64_from_iconst` with `i32_from_iconst` to avoid the `if-let`

Reply:
We can't do that because `i32_from_const` will unconditionally truncate upper bits, but we need to only match this rule when the `iconst` bits will losslessly fit in an `i32`.

Reply:
I don't think that's correct? The definition of `i32_from_iconst` uses `try_from`? (it very much should or we have a lot of incorrect lowering rules)

Reply:
`i32_from_iconst` seems to do the checks but then we have these other cases that do not:

```lisp
(decl u32_from_iconst (u32) Value)
(extractor (u32_from_iconst x)
           (u64_from_iconst (u64_as_u32 x)))

;; ...

(decl i8_from_iconst (i8) Value)
(extractor (i8_from_iconst x)
           (i32_from_iconst (i32_as_i8 x)))
```

Reply:
So I guess this particular one can change, but I don't think we can blanket apply this reasoning to the other cases